README.md
# zero-interaction-oidc-provider

This project provides an OIDC provider for use in development and testing.
The server will never require user interaction, making it perfect for automated browser testing.

[Read more about the motivation behind the project](./documentation/explanation/motivation.md).

## Usage

The server is available as a docker image.

```sh
docker run --rm \
    -p 9000:9000 \
    -e "REDIRECT_URI=http://localhost:4200/login/callback" \
    notclive/zero-interaction-oidc-provider
```

You should then provide the following configuration to your OIDC client:
* authority: http://localhost:9000
* client ID: zero-interaction-client-id

### Environment Variables

The simplest configuration can be accomplished using environment variables

| Variable | Description | Default |
| -------- | ----------- | ------- |
| REDIRECT_URI | The URI to be called after authentication, e.g. `http://localhost:4200/login/callback`. | `http://localhost:3000` |
| CLIENT_ID | The ID of your OIDC client. | `zero-interaction-client-id` |
| CLIENT_SECRET | The secret of your OIDC client, HTTP basic or POST authentication must be used when fetching tokens if and only if a secret is provided. | no secret |

### Behaviour

The server will always authenticate the user with a simple token containing a subject of fixed-user-id, and the openid scope.

CHANGELOG.md
## [1.2.0] - 2021-12-12

* Added a bundling step, using esbuild, when building the Docker image, to reduce the image size.

## [1.1.0] - 2021-11-19

* Changed how authorised session is injected, to avoid relying on cookies being stored

## [1.0.0] - 2021-11-14

* Initial public release


test/cypress/integration/default-token.spec.ts
describe('sign in with default token', () => {
    it('given provider has not been primed, when user signs in, then default token is issued', () => {
        // Given
        cy.visit('http://localhost:3000');

        // When
        cy.contains('button', 'Sign in').click();

        // Then
        cy.contains('Subject: fixed-account-id');
    });
});


src/index.ts
import {Provider} from 'oidc-provider';
import {Configuration, ConfigurationError, determineConfiguration, logConfiguration} from './configuration';
import {insertAuthorisedSession} from './insert-authorised-session';

try {
    const configuration = determineConfiguration();
    logConfiguration(configuration);
    const oidc = buildProvider(configuration);
    startServer(oidc, configuration.port);
} catch (error) {
    if (error instanceof ConfigurationError) {
        console.error(error.message);
        process.exitCode = 1;
    } else {
        throw error;
    }
}

function buildProvider(configuration: Configuration) {
    const oidc = new Provider(`http://localhost:${configuration.port}`, {
        clients: [{
            client_id: configuration.clientId,
            client_secret: configuration.clientSecret,
            redirect_uris: [configuration.redirectUri],
            // Expects basic or post authentication only if a client secret has been provided.
            token_endpoint_auth_method: configuration.clientSecret ? undefined : 'none'
        }],
        features: {
            // By default oidc-provider implements dummy login pages for development only.
            // We don't want to use these as they require interaction from the user.
            devInteractions: {
                enabled: false
            }
        },
        // Allow requests from any origin, this eliminates configuration at the cost of security.
        // Security isn't required here, this server is only for development and testing.
        clientBasedCORS: () => true
    });

    oidc.use(async function (context, next) {
        await insertAuthorisedSession(oidc, configuration.clientId, context.req);
        return next();
    });

    return oidc;
}

function startServer(oidc: Provider, port: number) {
    const server = oidc.listen(port, () => {
        console.log(`zero-interaction-oidc-provider listening at http://localhost:${port}`)
    });

    process.on('SIGINT', () => {
        console.log('Shutting down...');
        server.close();
    });
}

src/insert-authorised-session.ts
import * as http from 'http';
import {Provider} from 'oidc-provider';

const accountId = 'fixed-account-id';

export async function insertAuthorisedSession(
    oidc: Provider,
    clientId: string,
    request: http.IncomingMessage
) {
    const grantId = await generateGrant(oidc, clientId);
    const sessionId = await generateSession(oidc, clientId, grantId);
    insertSyntheticSessionCookie(oidc, request, sessionId);
}

async function generateGrant(oidc: Provider, clientId: string) {
    const grant = new oidc.Grant({clientId, accountId});
    grant.addOIDCScope('openid');
    return grant.save();
}

async function generateSession(oidc: Provider, clientId: string, grantId: string) {
    const session = new oidc.Session();
    session.accountId = accountId;
    session.grantIdFor(clientId, grantId);
    return session.save(60);
}

function insertSyntheticSessionCookie(oidc: Provider, request: http.IncomingMessage, sessionId: string) {
    const cookieName = (oidc as any).cookieName('session');
    request.headers.cookie = `${cookieName}=${sessionId}`;
}

src/configuration.ts
export interface Configuration {
    port: number;
    clientId: string;
    clientSecret?: string;
    redirectUri: string;
}

export function determineConfiguration() {
    return {
        port: determinePort() || 9000,
        clientId: process.env.CLIENT_ID || 'zero-interaction-client-id',
        clientSecret: process.env.CLIENT_SECRET,
        redirectUri: process.env.REDIRECT_URI || 'http://localhost:3000'
    };
}

function determinePort() {
    if (process.env.PORT === undefined) {
        return undefined;
    }
    if (Number.isInteger(process.env.PORT)) {
        return Number(process.env.PORT);
    }
    throw new ConfigurationError('Environment variable PORT must be numeric.')
}

export function logConfiguration(configuration: Configuration) {
    // Port is logged elsewhere.
    console.log('Starting zero-interaction-oidc-provider with the following client configuration:');
    console.log(` * client ID: ${configuration.clientId}`);
    console.log(` * client secret: ${describeClientSecret(configuration.clientSecret)}`);
    console.log(` * redirect URI: ${configuration.redirectUri}`);
}

function describeClientSecret(secret?: string) {
    if (secret === undefined) {
        return `none, authentication won't be required when fetching tokens`;
    }
    return `'${secret}', HTTP basic or POST authentication will be required when fetching tokens`;
}

export class ConfigurationError implements Error {

    private readonly _message: string;

    constructor(message: string) {
        this._message = message;
    }

    public get message() {
        return this._message;
    }

    public get name() {
        return 'Configuration error';
    }
}